tests/integration/cheatcodes/expect_revert/expect_revert_test.py

```diff
@@ -32,35 +37,21 @@ async def test_expect_revert(run_cairo_test_runner: RunCairoTestRunnerFixture):
             "test_with_except_revert_fail_expected",
             "test_fail_error_message",
         ],
+        expected_broken_test_cases_names=[
+            "test_already_expecting_error_message_when_no_arguments_were_provided"
+        ],
     )
 
 
-@pytest.fixture(autouse=True, scope="module", name="protostar")
-def protostar_fixture(create_protostar_project: CreateProtostarProjectFixture):
-    with create_protostar_project() as protostar:
-        yield protostar
-
-
-async def test_error_message_when_no_arguments_were_provided(
-    protostar: ProtostarFixture,
-):
-    (_, formatted_test_result) = await protostar.run_test_case("%{ expect_revert() %}")
-
-    assert "matching the following error" not in formatted_test_result
-    assert "Expected revert" in formatted_test_result
-
-
 async def test_already_expecting_error_message_when_no_arguments_were_provided(
-    protostar: ProtostarFixture,
+    testing_summary: TestingSummary,
 ):
-    (_, formatted_test_result) = await protostar.run_test_case(
-        """
-        %{
-            expect_revert()
-            expect_revert()
-        %}
-        """
+    (_, formatted_protostar_test_result) = get_protostar_test_case_result(
+        testing_summary,
+        "test_already_expecting_error_message_when_no_arguments_were_provided",
     )
 
-    assert "matching the following error" not in formatted_test_result
-    assert "Protostar is already expecting an exception." in formatted_test_result
+    assert "matching the following error" not in formatted_protostar_test_result
```

Comment:
what do you think about asserting fields of `BrokenTestCaseResult` (or so) here instead of looking for substrings in formatted representation? 

This way this test could be more precise, because such assertion is much stronger.

And the infra can be simpler, because without formatting phase, the `get_protostar_test_case_result` function can be easily replaced by `[]` operator implementation in TestingSummary (which has logical sense and thus I would love to see there).

Reply:
> what do you think about asserting fields of BrokenTestCaseResult (or so) here instead of looking for substrings in formatted representation?

This way this test could be more precise, because such assertion is much stronger.

I'd avoid testing internals. Checking test result type is done in test case above that uses `assert_cairo_test_cases`. 

> And the infra can be simpler, because without formatting phase, the get_protostar_test_case_result function can be easily replaced by [] operator implementation in TestingSummary (which has logical sense and thus I would love to see there).

- [ ] use `[]` operator

Reply:
> I'd avoid testing internals.

That's a good point, but I think `TestResult` structures are not so much internal aspects, given that we treat TestingSummary as tested entity and both are coupled.

Reply:
Testing `test_result.exception.message` doesn't guarantee printing the error to the user. Testing formatted output makes the test more useful. Such tests can be treated as [Subcutaneous Tests](https://martinfowler.com/bliki/SubcutaneousTest.html).

Checking the formatted output also checks the integration between `protostar.testing` and `protostar.commands`.

Reply:
I think that if you want to properly ensure that the feedback is done correctly, you would have to migrate this test to E2E. Right now, you assume that stringified result is printed.

On the other hand the following suite of tests is more lightweight:
1. Test that proper TestResult is returned for each case imaginable (that's the kind I push for here).
2. Test that abstractly constructed TestResults are stringified properly (that's part you try to mix in here).
3. Test on one example, in E2E suite, that indeed stringified representation is printed.

Unless I am very stupid, this set of tests should **definitely mathematically cover** all possible combinations of behaviours assuming sane implementation (no random, wild branches etc. crafted deliberately to break tests in the code).

Reply:
> Test that proper TestResult is returned for each case imaginable (that's the kind I push for here).

This is already handled by the test case above in `assert_cairo_test_cases`. I can add an extra assert in the new test case, but it's pointless.

> Test that abstractly constructed TestResults are stringified properly (that's part you try to mix in here).

This is done by "my" test case.

> Test on one example, in E2E suite, that indeed stringified representation is printed.

We have already plenty of E2E tests checking that formatted test results are indeed printed.

In summary, existing tests do all what you want.

---

> I think that if you want to properly ensure that the feedback is done correctly, you would have to migrate this test to E2E. Right now, you assume that stringified result is printed.

This remark conflicts with your suggestion about the suite of tests.

Reply:
Overloaded operator. Added an extra assert.

Reply:
Marek OOO. Resolving conversation.